File: scrape.js
// Select all product card elements
const productCards = document.querySelectorAll('.sf__pcard');

// Initialize an array to store the product data
const products = [];

// Loop through each product card to extract information
productCards.forEach(card => {
  const titleElement = card.querySelector('.sf__pcard-name');
  const imageElement = card.querySelector('.sf__image-box img');
  const descriptionElement = card.querySelector('.content-view-list .leading-7');
  
  const title = titleElement ? titleElement.innerText.trim() : 'No title';
  // Check for data-srcset attribute for lazy-loaded images
  const imageUrl = imageElement ? (imageElement.getAttribute('data-srcset') || imageElement.src).split(',')[0].split(' ')[0] : 'No image';
  const description = descriptionElement ? descriptionElement.innerText.trim() : 'No description';
  
  // Push the product data to the products array
  products.push({
    title: title,
    imageUrl: imageUrl,
    description: description
  });
});

// Convert the product data to JSON format
const jsonData = JSON.stringify(products, null, 2);

// Function to download data as a JSON file
function downloadJson(jsonData, filename) {
  const blob = new Blob([jsonData], { type: 'application/json' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

// Download the JSON file
downloadJson(jsonData, 'rolex_products.json');

// Log the product data to the console (optional)
console.log(products);
console.log(jsonData);

File: scrape.py
import json
import requests
from bs4 import BeautifulSoup
import time

# List of collections with their URLs, page counts, and output file names
collections = [
    {"url": "https://www.raretimeny.com/collections/rolex?page={}", "pages": 175, "file": "rolex.json"},
    {"url": "https://www.raretimeny.com/collections/audemars-piguet?page={}", "pages": 12, "file": "piguet.json"},
    {"url": "https://www.raretimeny.com/collections/patek-philippe?page={}", "pages": 13, "file": "patek.json"},
    {"url": "https://www.raretimeny.com/collections/cartier?page={}", "pages": 60, "file": "cartier.json"},
]

# Function to extract product information from a single page
def extract_product_info(soup):
    product_cards = soup.select('.sf__pcard')
    products = []
    for card in product_cards:
        title_element = card.select_one('.sf__pcard-name')
        image_element = card.select_one('.sf__image-box img')
        description_element = card.select_one('.content-view-list .leading-7')
        
        title = title_element.get_text(strip=True) if title_element else 'No title'
        
        image_url = 'No image'
        if image_element:
            if image_element.has_attr('data-src'):
                image_url = image_element['data-src'].split(',')[0].split(' ')[0]
            elif image_element.has_attr('src'):
                image_url = image_element['src']
        
        # Replace {width}x with 180x in the image URL if present
        image_url = image_url.replace("{width}x", "180x")
        
        description = description_element.get_text(strip=True) if description_element else 'No description'
        
        products.append({
            'title': title,
            'imageUrl': image_url,
            'description': description
        })
    return products

# Loop through each collection
for collection in collections:
    base_url = collection["url"]
    pages = collection["pages"]
    file_name = collection["file"]
    collection_products = []

    for page in range(1, pages + 1):
        print(f"Scraping page {page} of {base_url.format(page)}...")
        url = base_url.format(page)
        response = requests.get(url)
        if response.status_code == 200:
            soup = BeautifulSoup(response.content, 'html.parser')
            page_products = extract_product_info(soup)
            collection_products.extend(page_products)
        else:
            print(f"Failed to retrieve page {page}")
        time.sleep(1)  # Sleep for 1 second to avoid getting blocked

    # Save the product data to a JSON file
    with open(file_name, 'w') as json_file:
        json.dump(collection_products, json_file, indent=2)

    print(f"Scraped {len(collection_products)} products and saved to {file_name}")
